app/celery/tasks.py

```diff
@@ -874,3 +883,48 @@ def seed_bounce_rate_in_redis(service_id: str, interval: int = 24):
         bounce_rate_client.set_hard_bounce_seeded(service_id, bounce_data_dict)
 
     current_app.logger.info(f"Seeded hard bounce data for service {service_id} in Redis")
+
+
+@notify_celery.task(name="generate-report")
+@statsd(namespace="tasks")
+def generate_report(report_id: str):
+    current_app.logger.info(f"Generating report for Report ID {report_id}")
+    try:
+        report = get_report_by_id(report_id)
+
+        # mark the report as generating
+        report.status = ReportStatus.GENERATING.value
+        update_report(report)
+        # generate the report
+        url = create_report_in_s3(report)
+        report.url = url
+        report.generated_at = datetime.utcnow()
+        report.expires_at = datetime.utcnow() + timedelta(days=DAYS_BEFORE_REPORTS_EXPIRE)
+
+        # mark the report as ready
+        report.status = ReportStatus.READY.value
+        update_report(report)
+        current_app.logger.info(f"Report ID {str(report.id)} has been generated")
+    except Exception as e:
+        current_app.logger.exception(f"Failed to generate report for Report ID {report.id}: {str(e)}")
+        report.status = ReportStatus.ERROR.value
+        update_report(report)
+        raise
+
+
+def create_report_in_s3(report: Report) -> str:
+    """Creates a report in S3 and returns the URL"""
```

Comment:
maybe we should add some logging here to keep track of how long it is taking to generate this report?

Reply:
I'm seeing logs with timing for the `generate-report` task that calls this function - so that will give us the time to generate each report and the report id. I think the `@statsd` decorator is doing that for us. 
<img width="1150" alt="Screenshot 2025-04-01 at 1 24 00 PM" src="https://github.com/user-attachments/assets/6f421a2a-62b7-43d2-91c2-ba493a5db931" />